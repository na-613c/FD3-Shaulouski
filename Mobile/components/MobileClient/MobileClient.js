import React from 'react';
import PropTypes from 'prop-types';
import { mobileEvents } from '../../events';

import('./MobileClient.css');

class MobileClient extends React.PureComponent {

  static propTypes = {
    client: PropTypes.shape({
      id: PropTypes.number.isRequired,
      FIO: PropTypes.shape({
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
      }),
      balance: PropTypes.number.isRequired,
    })
  };

  state = {
    ...this.props.client
  };

  componentDidUpdate = (prevProps) => {
    const isNewProps = this.props.client.id !== prevProps.client.id ||
      this.props.client.balance !== prevProps.client.balance ||
      this.props.client.FIO.fam !== prevProps.client.FIO.fam ||
      this.props.client.FIO.im !== prevProps.client.FIO.im ||
      this.props.client.FIO.otch !== prevProps.client.FIO.otch;

    if (isNewProps) {
      this.setState({
        ...this.props.client,
        FIO: {
          ...this.props.client.FIO
        },
      });
    }
  };


  delete = (EO) => {
    mobileEvents.emit('EDelete', this.state.id);
  }

  edit = () => {
    mobileEvents.emit('EEdit', this.state);
  }

  render() {

    console.log("MobileClient id=" + this.props.client.id + " render");

    const isActive = this.state.balance > 0;

    return (
      <tr>
        <td>{this.state.FIO.fam}</td>
        <td>{this.state.FIO.im}</td>
        <td>{this.state.FIO.otch}</td>
        <td>{this.state.balance}</td>
        <td style={{ backgroundColor: isActive ? 'chartreuse' : 'red' }} >
          {isActive ? 'Active' : 'Blocked'}
        </td>
        <td><button className='MobileClient-Edit' onClick={this.edit} >Редактировать</button></td>
        <td><button className='MobileClient-Delete' onClick={this.delete}>Удалить</button></td>
      </tr>
    )
  }
}

export default MobileClient;
