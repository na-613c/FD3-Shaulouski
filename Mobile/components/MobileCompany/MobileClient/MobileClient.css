div.MobileClient {
  padding: 2px 0;
}

span.MobileClientBalance {
  float: right;
  width: 40px;
  font-weight: bold;
  text-align: right;
}

span.MobileClientFIO {
  margin-right: 30px;
}