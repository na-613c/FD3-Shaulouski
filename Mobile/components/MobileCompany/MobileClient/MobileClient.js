import React from 'react';
import PropTypes from 'prop-types';
import { mobileEvents } from '../../../events';

import './MobileClient.css';

class MobileClient extends React.PureComponent {

  static propTypes = {
    client: PropTypes.shape({
      id: PropTypes.number.isRequired,
      FIO: PropTypes.shape({
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
      }),
      balance: PropTypes.number.isRequired,
    })
  };

  state = {
    ...this.props.client
  };

  componentWillReceiveProps = (newProps) => {
    const isNewProps = this.state.id !== newProps.client.id ||
      this.state.balance !== newProps.client.balance ||
      this.state.FIO.fam !== newProps.client.FIO.fam ||
      this.state.FIO.im !== newProps.client.FIO.im ||
      this.state.FIO.otch !== newProps.client.FIO.otch;

    if (isNewProps) {
      this.setState({
        ...newProps.client,
        FIO: {
          ...newProps.client.FIO
        },
      });
    }
  };



  delete = (EO) => {
    mobileEvents.emit('EDelete', this.state.id);
  }

  edit = () => {
    mobileEvents.emit('EEdit', this.state);
  }

  render() {

    console.log("MobileClient id=" + this.props.client.id + " render");

    const isActive = this.state.balance > 0;

    return (
      <tr>
        <td>{this.state.FIO.fam}</td>
        <td>{this.state.FIO.im}</td>
        <td>{this.state.FIO.otch}</td>
        <td>{this.state.balance}</td>
        <td style={{ backgroundColor: isActive ? 'chartreuse' : 'red' }} >
          {isActive ? 'Active' : 'Blocked'}
        </td>
        <td><button onClick={this.edit} >Редактировать</button></td>
        <td><button onClick={this.delete}>Удалить</button></td>
      </tr>
    )

  }

}

export default MobileClient;
