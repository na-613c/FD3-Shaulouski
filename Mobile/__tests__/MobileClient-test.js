import React from 'react';
import renderer from 'react-test-renderer';

import MobileClient from '../components/MobileClient/MobileClient';

test('работа MobileClient', () => {

    const client = { id: 101, FIO: { fam: "Иванов", im: "Иван", otch: "Иванович" }, balance: 200 };

    const component = renderer.create(
        <MobileClient client={client} />
    );

    // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();

    // найдём в вёрстке компонента саму кнопку
    const buttonEdit = component.root.findByProps({ className: "MobileClient-Edit" })
    const buttonDelete = component.root.findByProps({ className: "MobileClient-Delete" })

    // и "нажмём" на неё
    buttonEdit.props.onClick();
    buttonDelete.props.onClick();

    // получаем уже изменённый снэпшот
    componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();

    // "нажмём" кнопку ещё раз
    buttonEdit.props.onClick();
    buttonDelete.props.onClick();

    // и получаем окончательный снэпшот
    componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();

    /*
    // можно эмулировать события, передавая в качестве объекта события то что нам нужно:
    wrapper.find('select').simulate('change', {
        target: { value: "hello" },
    });
    */

});