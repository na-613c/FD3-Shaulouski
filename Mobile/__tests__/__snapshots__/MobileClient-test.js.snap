// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа MobileClient 1`] = `
<tr>
  <td>
    Иванов
  </td>
  <td>
    Иван
  </td>
  <td>
    Иванович
  </td>
  <td>
    200
  </td>
  <td
    style={
      Object {
        "backgroundColor": "chartreuse",
      }
    }
  >
    Active
  </td>
  <td>
    <button
      className="MobileClient-Edit"
      onClick={[Function]}
    >
      Редактировать
    </button>
  </td>
  <td>
    <button
      className="MobileClient-Delete"
      onClick={[Function]}
    >
      Удалить
    </button>
  </td>
</tr>
`;

exports[`работа MobileClient 2`] = `
<tr>
  <td>
    Иванов
  </td>
  <td>
    Иван
  </td>
  <td>
    Иванович
  </td>
  <td>
    200
  </td>
  <td
    style={
      Object {
        "backgroundColor": "chartreuse",
      }
    }
  >
    Active
  </td>
  <td>
    <button
      className="MobileClient-Edit"
      onClick={[Function]}
    >
      Редактировать
    </button>
  </td>
  <td>
    <button
      className="MobileClient-Delete"
      onClick={[Function]}
    >
      Удалить
    </button>
  </td>
</tr>
`;

exports[`работа MobileClient 3`] = `
<tr>
  <td>
    Иванов
  </td>
  <td>
    Иван
  </td>
  <td>
    Иванович
  </td>
  <td>
    200
  </td>
  <td
    style={
      Object {
        "backgroundColor": "chartreuse",
      }
    }
  >
    Active
  </td>
  <td>
    <button
      className="MobileClient-Edit"
      onClick={[Function]}
    >
      Редактировать
    </button>
  </td>
  <td>
    <button
      className="MobileClient-Delete"
      onClick={[Function]}
    >
      Удалить
    </button>
  </td>
</tr>
`;
